core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MergeProjectWithValues.java

```diff
@@ -0,0 +1,201 @@
+        if (valuesNode.getOutputSymbols().isEmpty()) {
+            ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+            for (int i = 0; i < valuesNode.getRowCount(); i++) {
+                projectedRows.add(new Row(ImmutableList.copyOf(assignments.values())));
+            }
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), node.getOutputSymbols(), projectedRows.build()));
+        }
+
+        // do not proceed if ValuesNode contains a non-deterministic expression and it is referenced more than once by the projection
+        Set<Symbol> nonDeterministicValuesOutputs = new HashSet<>();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                if (!isDeterministic(row.getItems().get(i), metadata)) {
+                    nonDeterministicValuesOutputs.add(valuesNode.getOutputSymbols().get(i));
+                }
+            }
+        }
+        Set<Symbol> multipleReferencedSymbols = node.getAssignments().getExpressions().stream()
+                .map(SymbolsExtractor::extractAll)
+                .flatMap(Collection::stream)
+                .collect(toImmutableMultiset()).entrySet().stream()
+                .filter(entry -> entry.getCount() > 1)
+                .map(Multiset.Entry::getElement)
+                .collect(toImmutableSet());
+        if (!Sets.intersection(nonDeterministicValuesOutputs, multipleReferencedSymbols).isEmpty()) {
+            return Result.empty();
+        }
+
+        // inline values expressions into projection's assignments
+        RowType projectedRowType = RowType.anonymous(assignments.keySet().stream()
+                .map(symbol -> context.getSymbolAllocator().getTypes().get(symbol))
+                .collect(toImmutableList()));
+        ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            ImmutableMap.Builder<SymbolReference, Expression> mapping = ImmutableMap.builder();
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < row.getItems().size(); i++) {
+                mapping.put(valuesNode.getOutputSymbols().get(i).toSymbolReference(), row.getItems().get(i));
+            }
+            Row projectedRow = new Row(assignments.values().stream()
+                    .map(expression -> replaceExpression(expression, mapping.build()))
+                    .collect(toImmutableList()));
+
+            // do not proceed if the resulting expression has symbols (correlation)
+            if (!SymbolsExtractor.extractUnique(projectedRow).isEmpty()) {
+                return Result.empty();
+            }
+
+            // do not inline expressions which cannot be evaluated by ExpressionInterpreter
+            // otherwise, they will cause failure in execution planner
```

Comment:
What's an example of this? The only think I can think of is an expression that contains a reference to a column from an outer query. But that should never make it to the execution engine.

Reply:
There's a test case which covers it `TestMergeProjectWithValues.testExpressionNotInterpreted()`.
The expression used there is `IF(random() > 2, 1, 0)`.

Reply:
Isn’t that covered by the non-determinism check above?

Reply:
No, the non-determinism check works the other way:
if there's a non-deterministic expression in ValuesNode, we mustn't inline it twice (or more) in the ProjectNode's expression.
In this test case, there's a non-deterministic projection over deterministic Values.

Reply:
Ok, then I’d like to understand why it’s failing later on. While the expression optimizer shouldn’t actualize non-deterministic expressions, the vanilla interpreter should. In this case the non-deterministic expression should be actualized for each of the entries in the ValuesNode as an eager evaluation of the table defined by it.

Reply:
I'll try to investigate why the interpreter won't interpret this expression and let you know.
But anyway, I think it's a good move to make sure that we don't put into ValuesNode an expression that would fail at executionPlanner.

Reply:
That’s what I’m referring to... it shouldn’t fail in executionPlanner for a ValuesNode, which is effectively a constant at that point